### ISSUE TITLE
Calling pthread_join in uffd monitor stop function causes deadlock with multiple monitors

### ISSUE THREAD

**Issue by wckzhang** [ofiwg/libfabric#7003]
Using multiple monitors caused uffd's stop funtion to deadlock in the pthread_join function when using an import monitor (due to the import monitor notifying and trying to obtain a read lock which it will never get). If we expect to support multiple monitors, then the stop function should not have any code that could result in a free, so we probably need to delay pthread_join into the cleanup phase.

**Comment by shefty** [comment]
Stop is the cleanup phase for the monitor.  I'm assuming stop is being called from ofi_monitors_del_cache() <- ofi_mr_cache_cleanup(), so the cache is going away.

I don't see an obvious solution to this.

**Comment by wckzhang** [comment]
Each monitor has a .cleanup function, if stop is meant to be cleanup why do we have a cleanup function?

**Comment by shefty** [comment]
Hmm... why _do_ we have a cleanup call?  AFAICT, all the cleanup calls point to ofi_monitor_cleanup(), which doesn't do anything other than assert a list is empty.  I guess cleanup pairs with init, which only initializes the list.  Unless I'm missing something in the implementation, init/cleanup could easily be merged into start/stop.

But that won't solve the problem here.  We might be able to move calling stop() outside of the mm_list_rwlock in ofi_monitors_del_cache().  That could be racy, and I'm not sure the mm_list_rwlock is the lock causing issues.  Do you know which lock is trying to be acquired?

**Comment by shefty** [comment]
The problem with the current code is only if there are multiple monitors for the same hmem type, correct?

**Comment by wckzhang** [comment]
Yeah, this was exposed with opening the import monitor while manually setting the uffd monitor

**Comment by wckzhang** [comment]
Oh and yeah it's the rwlock, what's happening is the writer lock is acquired during the stop phase, but pthread_join kicks off memhooks which tries to grab a read lock it will not acquire.

**Comment by wckzhang** [comment]
I can write some new code to change uffd monitor's cleanup function to contain the pthread_join. Let me know if you have any objections Sean, I'll go test it now

**Comment by shefty** [comment]
Aren't init/cleanup called once, but start/stop may be called repeatedly? That's my main concern with trying to move join into cleanup, versus finding a way to move it outside of the lock but in stop.

**Comment by wckzhang** [comment]
Hmm..I'll take a look at the code more today

**Comment by wckzhang** [comment]
Did some investigating...

I think calling notify during the ofi_monitors_del_cache cleanup phase will probably be harmless, it doesn't seem like the notify path will attempt to grab a lock that has been closed and the rbmap isn't deleted until after this phase. Theoretically the tree should be empty so we shouldn't enter the util_mr_uncache_entry function.

I think it should be possible to move the uffd start and stop code into uffd init and uffd cleanup respectively. Then start/stop can be a no-op. I don't see any reason this can't happen. I actually think this *should* happen. Since the uffd monitor is a global object, calling start multiple times makes no sense on it and it's possible that there may be a bug with it living in the start code.

Other than that, the question is whether we should move the monitor stopping outside of the rwlock. From what I can tell (since I forgot the original reason I wrote the rwlock), it's supposed to protect the dlist_entry list of ofi_mem_monitor. If we had the monitor->stop code outside of this lock, then what could happen is someone calls add cache, and then we promptly stop the monitor resulting in a stopped monitor for the cache. This scenario seems unlikely, but I don't see how we can move the stop function outside the rwlock. Though it may be possible to try to obtain a read lock on cleanup instead of a write lock. Having a read lock on cleanup will prevent this deadlock from happening and will also protect the del_cache code from executing at the same time as add_cache.

So I think we need to do two things:

1. Move uffd start/stop into init/cleanup <-- I think this is correct since the functions access the global uffd object
2. Change the rwlock in ofi_monitors_del_cache to be a read lock instead of a write lock.

What do you think Sean?

**Comment by wckzhang** [comment]
Thinking a little more, dlist_remove(&cache->notify_entries[iface]); is located in the cleanup code. If we're using a reader lock, this dlist_remove could occur at the same time as dlist_foreach_container if a notify occurs. I don't think we can let this happen so maybe the reader lock isn't possible.

**Comment by shefty** [comment]
The reason for start/stop being separate is so that we don't start the uffd thread unless it's actually needed by some provider.  If the app is only using the tcp provider, uffd should never be enabled, even if there are providers present that need the cache.  Likewise, once the app is done with a provider, any associated cache can go away.  Init/cleanup are general calls invoked at library load/unload.

The same is true for memhooks.  If we don't need the cache because the app only intends to use the tcp provider, we don't want to intercept the memory calls.

We may have more flexibility on the stop call.  Once we start a cache, we _could_ keep it active until library unload.  Some developer is going to complain about this someday -- that once we no longer need the resources for some object, we should release them.  But folding stop into cleanup is an option that most apps aren't going to care about.

We can solve this as long as we can move the problematic code outside of the rwlock, even if the functionality remains in the stop call.  If we can guarantee that start won't be called while stop is running, we should be race free.  A new start/stop lock could protect against that.

**Comment by wckzhang** [comment]
I don't think that just guaranteeing start won't be called while stop is running will be race free, we can do that by changing the stop's writer lock to a reader lock. The problem is that we also need notify's not to happen since the dlist walking cannot occur during start or stop.

What you said about start/stop makes sense, I suppose we could add a counter to prevent this from happening multiple times. Maybe we can move the uffd stop code into cleanup by itself? It's kind of janky, but I'm not seeing a great way of handling the deadlock.

**Comment by wckzhang** [comment]
Okay yeah, I get what you're saying now. It looks like that would work, let me write something up, test, and PR

**Comment by shefty** [comment]
Fixed by #7063
